docs: Added guide to migrate to Ubuntu 24.04 (#409)
elastisys/compliantkubernetes-kubespray: migration/ubuntu/20-to-24.md

Reviewer: Please rename this file to `20.04-to-22.04-openstack.md` to make it clearer that it is only for openstack.

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/064086c3fc34aae568490aa93c8d77bcce016df3

Reviewer: Yeah, that was an error on my part of course it should be `20.04-to-24.04-openstack.md` :sweat_smile:

Author: Yea, I should've seen that as well :see_no_evil:

# Migrate to Ubuntu 24.04 on Openstack

This migration assumes that the environment is running on an Openstack environment.
This mainly applies to the terraform details, so you should be able to follow along with the kubespray steps.

> 

Reviewer 2: Shall we add a note on that about 50% openstack qoutas needs to be available?

Reviewer 2: Maybe a `Prerequisites` header?

Author: Makes sense, I guess either that or that you do one node at a time which will work but will take significantly more time

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/e79e915dc4ade6c119818be3ec1eb7e14e25f700

  kubectl -n kube-system exec -it "etcd-${HOSTNAME}" -- \
    etcdctl --endpoints=https://127.0.0.1:2379 \
            --cacert=/etc/kubernetes/ssl/etcd/ca.crt \
            --cert=/etc/kubernetes/ssl/etcd/server.crt \
            --key=/etc/kubernetes/ssl/etcd/server.key \
            member list
  ```

- If there are old nodes left, remove them using the `uuid` from the previous command (The id in the first column of each row)

  ```bash
  HOSTNAME="$(kubectl get node -l node-role.kubernetes.io/control-plane= -ojsonpath='{.items[0].metadata.name}')"
  kubectl -n kube-system exec -it "etcd-${HOSTNAME}" -- \
    etcdctl --endpoints=https://127.0.0.1:2379 \
            --cacert=/etc/kubernetes/ssl/etcd/ca.crt \
            --cert=/etc/kubernetes/ssl/etcd/server.crt \
            --key=/etc/kubernetes/ssl/etcd/server.key \
            member remove <uuid>
  ```

- Check that old control plane nodes are removed:

  ```bash
  kubectl get nodes
  ```

- If there are any old nodes left, delete them

  ```bash
  kubectl delete nodes <node>,...
  ```

- Change back to using dynamic inventory

  ```bash
  pushd "${CK8S_CONFIG_PATH}"

>   cp -r <sc|wc>-config-backup/inventory.ini <sc|wc>-config/inventory.ini

Reviewer: Do we need recursive here?

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/9bf216c10273ee3ba15810b029a3366a796a7e94

  popd
  ```

- Remove old control plane nodes using Terraform
- Remove the old control plane floating IPs from group vars under the key `supplementary_addresses_in_ssl_keys`

> [!NOTE]
> This is not needed if you are using a loadbalancer in front of the API server.

- Apply kubespray:

  ```bash
  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
  ck8s-kubespray apply <sc|wc> --tags=master,nginx -b
  popd
  ```

- Update the NetworkPolicies:

  ```bash
  pushd "${CK8S_APPS_REPOSITORY_PATH}"
  ck8s update-ips <sc|wc> apply

  ck8s ops helmfile <sc|wc> -l app=netpol -i apply
  popd
  ```

### Replace worker nodes

- Add a new worker node that doesn't specify `image_id` (To use the new image)
- Verify by running `terraform plan` and make sure there's a new node added with the correct image
- Add the rest of the new worker nodes
- Run `terraform apply` to add all the new nodes
- Add all the new worker nodes

  ```bash
  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
  ck8s-kubespray run-playbook <sc|wc> facts.yml
  ck8s-kubespray run-playbook <sc|wc> scale.yml -b --limit=<new-node1>,<new-node2>,...
  popd
  ```

> - Drain the old nodes and make sure everything starts up properly

Reviewer: We will need to make some considerations here for environments hosting applications with allowlisted consumers, as this will change IPs.

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/893a1ead1b3e07cdc1bd7b4ba6ca32ca0aa2ba0f

# Migrate to Ubuntu 24.04 on Openstack

> This migration assumes that the environment is running on an Openstack environment.
> This mainly applies to the terraform details, so you should be able to follow along with the kubespray steps.

Reviewer: ```suggestion
> [!important]
> This migration is written for environments running on OpenStack.
> This mainly applies to the Terraform details, so you should be able to adapt those and follow along with the Kubespray steps for other infrastructure providers.
```

> This mainly applies to the terraform details, so you should be able to follow along with the kubespray steps.

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/ac762a4ee1438cd18875d2979e8fcbff1b1cf23e

## Prerequisites

Before you start you should make sure that if you plan to follow this guide and create all new nodes at the same time, do make sure that you have enough quota for that.

> Recommended is to do at least all the control planes for one cluster in one go, but for worker nodes you can replace in smaller groups at the time if it's a large cluster.
> 
> ## Prepare

Reviewer: ```suggestion
Recommended is to do at least all the control planes for one cluster in one go, but for worker nodes you can replace in smaller groups at the time if it's a large cluster.

_Warning about that nodes will change ips so there needs to be communication regarding this._

## Prepare
```

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/ac762a4ee1438cd18875d2979e8fcbff1b1cf23e

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/ac762a4ee1438cd18875d2979e8fcbff1b1cf23e

These steps can be done without any disruption to the running cluster.

- Prepare environment variables

  ```bash
  export CK8S_CONFIG_PATH="/path/to/cluster/config"
  export CK8S_KUBESPRAY_REPOSITORY_PATH="/path/to/compliantkubernetes-kubespray"
  export CK8S_APPS_REPOSITORY_PATH="/path/to/compliantkubernetes-apps"
  export KUBECONFIG="${CK8S_CONFIG_PATH}/.state/kube_config_<sc|wc>.yaml"
  ```

- In `${CK8S_CONFIG_PATH}/<sc|wc>-config/cluster.tfvars` update the following:

  - Change global `image_uuid` to the ID of Ubuntu 24.04

>   - Change all existing nodes and add `image_id` to be the old image id

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/ac762a4ee1438cd18875d2979e8fcbff1b1cf23e

- Verify by running `terraform plan` and make sure there's no changes on any on the existing nodes
- Add a new control plane node that doesn't specify `image_id` (To use the new image)
- Verify by running `terraform plan` and make sure there's a new node added with the correct image
- Add the rest of the new control plane nodes
- Run `terraform apply` to add all the new nodes
- Backup the current state:

  ```bash
  pushd "${CK8S_CONFIG_PATH}"
  cp -r <sc|wc>-config <sc|wc>-config-backup
  popd
  ```

- Generate new inventory:

  ```bash
  ./gen-inventory.sh <sc|wc>
  ```

- Reorder inventory so the old control plane nodes are first in each category they occur in
- Add the new control plane floating IPs into group vars under the key `supplementary_addresses_in_ssl_keys`

> [!NOTE]

> > This is not needed if you are using a loadbalancer in front of the API server or only use DNS records.

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/ac762a4ee1438cd18875d2979e8fcbff1b1cf23e

  popd
  ```

- Remove old control plane nodes using Terraform
- Remove the old control plane floating IPs from group vars under the key `supplementary_addresses_in_ssl_keys`

> [!NOTE]
> This is not needed if you are using a loadbalancer in front of the API server.

- Apply kubespray:

  ```bash
  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
  ck8s-kubespray apply <sc|wc> --tags=master,nginx -b
  popd
  ```

- Update the NetworkPolicies:

  ```bash
  pushd "${CK8S_APPS_REPOSITORY_PATH}"
  ck8s update-ips <sc|wc> apply

  ck8s ops helmfile <sc|wc> -l app=netpol -i apply
  popd
  ```

### Replace worker nodes

- Add a new worker node that doesn't specify `image_id` (To use the new image)
- Verify by running `terraform plan` and make sure there's a new node added with the correct image
- Add the rest of the new worker nodes
- Run `terraform apply` to add all the new nodes
- Add all the new worker nodes

  ```bash
  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
  ck8s-kubespray run-playbook <sc|wc> facts.yml
  ck8s-kubespray run-playbook <sc|wc> scale.yml -b --limit=<new-node1>,<new-node2>,...
  popd
  ```

- Drain the old nodes and make sure everything starts up properly

> > [!WARNING]
> > If you are using applications that allowlist the old nodes, please update that list to also include the new nodes before draining the old ones

Reviewer: ```suggestion
> [!caution]
> If you are using applications that allowlist the old nodes, please update that list to also include the new nodes before draining the old ones.
```

Make this harder because here comes the impact :slightly_smiling_face:

Author: https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/ac762a4ee1438cd18875d2979e8fcbff1b1cf23e
